File: Python3/0290_word_pattern.py
# https://leetcode.cn/problems/word-pattern/

class Solution:
    def wordPattern(self, pattern, s):
        """
        Inputs:
            pattern: str
            s: str
        Outputs:
            res: bool
        """
        words = s.split()
        if len(pattern) != len(words):
            return False

        hash_ps = dict()
        hash_sp = dict()
        for i in range(len(pattern)):
            p = pattern[i]
            word = words[i]
            if hash_ps.get(p, -1) == -1 and hash_sp.get(word, -1) == -1:
                hash_ps[p] = word
                hash_sp[word] = p
            
            elif hash_ps.get(p, -1) != word or hash_sp.get(word, -1) != p:
                return False

        return True

File: Cpp/0290_word_pattern.cpp
// https://leetcode.cn/problems/word-pattern/

class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> words = split(s);
        if (pattern.size() != words.size()) {
            return false;
        }

        unordered_map<char, string> hash_ps;
        unordered_map<string, char> hash_sp;
        for (int i = 0; i < pattern.size(); ++i) {
            char p = pattern[i];
            string word = words[i];

            if ((hash_ps.find(p) == hash_ps.end()) &&
                ((hash_sp.find(word) == hash_sp.end()))) {
                hash_ps[p] = word;
                hash_sp[word] = p;
            }
            else if ((hash_ps.find(p) != hash_ps.end()) &&
                (hash_sp.find(word) != hash_sp.end())) {
                if ((hash_ps[p] != word) || (hash_sp[word] != p)) {
                    return false;
                }
            }
            else {
                return false;
            }
        }
        return true;
    }

    vector<string> split(string s) {
        vector<string> res;
        string word;
        for (char c: s) {
            if (c == ' ') {
                res.push_back(word);
                word.clear();
            }
            else {
                word = word + c;
            }
        }
        res.push_back(word);
        return res;
    }
};
